src/ublox.cpp

```diff
@@ -206,50 +255,143 @@ void UBLOX::read_cb(uint8_t byte)
+      if (new_data())
+        last_valid_message_ = millis();
     }
     else
     {
       // indicate error if it didn't work
       num_errors_++;
+      DBG("failed to parse message\n");
       parse_state_ = START;
     }
   }
 
   prev_byte_ = byte;
 }
 
-void UBLOX::get_pos_ecef(double* pos_ecef, uint32_t* t_ms)
+// convert a time struct to unix time
+// dumb leap years
+// This function will break in 2100, because it doesn't take into account that 2100 is not a leap year
+uint64_t convert_to_unix_time(const UBLOX::GNSS_TIME_T &time)
 {
-  (void)t_ms;
-  pos_ecef[0] = pos_ecef_message_.ecefX;
-  pos_ecef[1] = pos_ecef_message_.ecefY;
-  pos_ecef[2] = pos_ecef_message_.ecefZ;
+  uint32_t day_s = 24 * 60 * 60; // seconds per day
+  uint32_t year_s = 365 * day_s; // seconds per non-leap day
+  uint32_t elapsed_years = time.year - 1970;
+  uint32_t elapsed_leap_days = (elapsed_years + 1) / 4;
+  if (time.year % 4 == 0) // If currently in a leap year
+    if (time.month >= 3)  // If past feb 29
+      elapsed_leap_days++;
+  uint32_t elapsed_days; // Days past in the year
+  switch (time.month)
+  {
+  case 1:
+    elapsed_days = 0;
+    break;
+  case 2:
+    elapsed_days = 31;
+    break;
+  case 3:
+    // Ignore leap days, because it is accounted for above
+    elapsed_days = 31 + 28;
+    break;
+  case 4:
+    elapsed_days = 31 + 28 + 31;
+    break;
+  case 5:
+    elapsed_days = 31 + 28 + 31 + 30;
+    break;
+  case 6:
+    elapsed_days = 31 + 28 + 31 + 30 + 31;
+    break;
+  case 7:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30;
+    break;
+  case 8:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30 + 31;
+    break;
+  case 9:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30 + 31 + 31;
+    break;
+  case 10:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30;
+    break;
+  case 11:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31;
+    break;
+  case 12:
+    elapsed_days = 31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31 + 30;
+    break;
+  default: // This should never be reached, because there are 12 month / year
+    elapsed_days = 0;
+  }
+  elapsed_days += time.day - 1; // Minus 1 because the day is not yet complete
+  return elapsed_years * year_s + (elapsed_days + elapsed_leap_days) * day_s + (time.hour * 60 + time.min) * 60
+         + time.sec;
+}
+/* Tells if new data is available
+ * Only returns true if new data has been recieved, and if all three sources match time of week.
+ * This is because if the times of week do not match, new data is still being recieved,
+ * and attempting to read could result in data from different times.
+ */
+bool UBLOX::new_data()
+{
+  return this->new_data_ && (this->nav_message_.iTOW == this->pos_ecef_.iTOW)
+         && (this->nav_message_.iTOW == this->vel_ecef_.iTOW);
+}
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers" // Ignore warning about leaving struct fields blank
```

Comment:
I don't like the pattern of 

`T data = {};`
and then filling in every value.  First off, it means that we have this horrible GCC hack
second, it hides from us uninitialized data problems
third, it's extra slow.

I would be in favor of removing the zero-initialization step for each of these data structures

Reply:
I removed this from ublox.cpp. The fields in ublox.h are still initialized with empty structs in ublox.h, so that they are zeroed instead of reading garbage data.